### DISCUSSION TITLE
[Cluster][Update moose][Error in app]

### DISCUSSION THREAD

**Discussion by coskrrb2002** [idaholab/moose#17568] [Q&A Getting Started] [answered]
Dear moose users,

I am using a linux cluster and recently I have updated a moose in the cluster.

But, eventually it shows an error while I am compiling my app after updating the moose.

I updated the moose in the following steps.

I updated both libmesh and petsc.

Then, I updated moose with git fetch.

After that I recompile my app with

make clobberall

make -j20

But in make command, I got the following error.

![image](https://user-images.githubusercontent.com/35205601/114292355-e1f50500-9ac8-11eb-9da3-4b75c39b8fc5.png)

I guess this error is related with library...

But I don't know which library should I have to update.

Kind regards

**Comment by GiudGiud** [comment]
Hello

Did you only run git fetch to update moose? fetching isnt enough, you need to git pull as well.
Alternatively you can check out the latest commit with `git checkout origin master`. 

Just to be sure, which moose commit are you currently using? 

Best,
Guillaume

**Reply by coskrrb2002** [reply]
I did the git pull as well, but it didn't work.

I did git checkout origin master and got the following result:
![image](https://user-images.githubusercontent.com/35205601/114342912-c0af1a00-9b97-11eb-8923-e8da99445d9c.png)

Its very ridiculous, cause in my personal computer, my app works fine with moose after the updating.

**Reply by GiudGiud** [reply]
Ok this is the correct commit for the latest moose. 
The git pull did not work, or you got the error pasted initially with git fetch/pull ?

Tagging @loganharbour, he will know what's up with this error in building the ray tracing module

**Reply by coskrrb2002** [reply]
Thank you,

git pull said everything is up to date.

So, I guess my moose is the lastest version one.

Except the ray tracing module, other modules works fine.

**Reply by roystgnr** [reply]
@loganharbour asked me to take a look, and at first glance it looks like somehow C++14-specific code is in that module, but you're trying to compile using a C++11-only compiler.  (a ten year old standard, but one which we're still trying to support! no idea how our CI let C++14-requiring code in!)

He's now trying to rewrite that code in a C++11-compatible way, which I hope will be possible but which might take some serious work, so in the meantime you might see if you can upgrade your compiler as a workaround?  And if you *can't* easily get your hands on a compiler to handle this, then I'd love to know about it: I was hoping to update the libMesh C++ requirements, possibly jumping straight to C++17, late this year, but I can hit the brakes on that plan if there are still a number of downstream users stuck with too-old compilers.

BTW, in the future, could you paste compiler error/warning messages
```
like this,
```
which in Github markdown can be formatted
~~~
``` 
like this,
``` 
~~~
rather than using screenshots?  Screenshots don't make it into search engines and so aren't as helpful if future users hit similar problems.

**Comment by loganharbour** [comment, marked as answer]
This issue will be resolved in #17582. See an explanation in https://github.com/idaholab/moose/discussions/17568#discussioncomment-600418.

However, we've discussed in the past that we loosely require c++14. #17581 enforces that requirement - we'll see if we can get it in or not.

@coskrrb2002 I'm happy to help you with getting MOOSE compiling in the interim, before either of the PRs that I mentioned above ar merged. We're in the middle of some system downtime, so they won't be merged quickly. Are you using conda for MOOSE, or did you compile libMesh on your own using `scripts/update_and_rebuild_libmesh.sh`?

**Reply by coskrrb2002** [reply]
Dear @loganharbour,

I am not using conda for MOOSE in my cluster. So, I just compiled libMesh by ./scripts/update_and_rebuild_libmesh.sh

Thank you for your explanation!

**Comment by WilkAndy** [comment]
So we need C++14.  Just for our education, can you explain (or point to doco/stackexchange) the relevant change in C++14 ?

**Reply by roystgnr** [reply]
Take a look at the restrictions on constexpr functions in the bullet points tagged "until C++14" [about a quarter of the way down this page](https://en.cppreference.com/w/cpp/language/constexpr).  You basically only get to have a single statement, your return statement; any function that you can't write as a one-liner can't be constexpr.

That said, I'm pretty sure the capability here *can* be written in terms of constexpr one-liners, so I wouldn't say we "need" C++14 so much as we "really really want" C++14; the rewrite won't be impossible, it'll just be a big pain in the neck.

Plus, this is just the latest addition to the list of reasons users have had for wanting to upgrade.  INL folks [started asking me about support last year](https://github.com/libMesh/libmesh/issues/2635).  I've been using C++14 in a few small projects since before it was an official standard; for anybody trying to use expression template classes there's just no decent substitute for C++14-or-better auto return type deduction.
